posts/coding-intentionally-in-bash-grains.md

```diff
@@ -0,0 +1,185 @@
+
+```bash
+bc <<< "ibase=2;1111111111111111111111111111111111111111111111111111111111111111"
+```
+
+More intentional, because it more closely matches what the problem gives us.  But, with it being so long and with us not speaking robot, a long, essentially uncountable string of 1's is maybe not an improvement.
+
+### Second option: the brute-force calculation
+
+OK, so maybe we abandon the non-decimal numbering systems altogether.  Why don't we make the code match how we would total up the number of grains on a chessboard by hand, by counting the grains on each square?
+
+```bash
+total=0
+current_grains=1
+
+for square in {1..64}; do
+  total=$( bc <<< "$total + $current_grains" )
+  current_grains=$( bc <<< "$current_grains * 2" )
+done
+
+echo "$total"
+```
+
+That's a lot more readable and understandable.  The code clearly shows that the number of squares on the chess board is a driving factor, as well as the doubling effect for each square.  I think this is better than the initial solution.
+
+However.
+
+It's slow.  Looping, adding, and repeatedly calling out to an external command?  They all stack up to a slow-ish run-time.  Now, is that a huge deal?  No.  If you're scripting this in Bash, you've probably already decided you don't have speed constraints.  But could it be better?  Yes.
+
+### Third option: direct calculation
+
+In the math biz (that's what cool math folks call it), we've got a name for the process "take something and then multiply it by the same number a bunch of times": exponents.
+
+In fact, the number of grains on each square can be expressed as:
+
+```bash
+bc <<< "2^($square - 1)"
+# We subtract one from $square because square 1 has 1 (2^0) grains on it.
+```
+
+So, how do we add all these up without iterating?
+
+Let's consider a *smaller* version of the same problem: a chessboard with 5 squares[^3].
+
+The five squares would have the following number of grains:
+
+```txt
+---------------------
+| 1 | 2 | 4 | 8 |16 |
+---------------------
+```
+
+And the total here would be: 1 + 2 + 4 + 8 + 16 = 31.  Hm.  31 doesn't scream anything obvious at me yet.  Let's go a little bigger.
+
+OK, well what about a 6-square chessboard?
+
+```txt
+-------------------------
+| 1 | 2 | 4 | 8 |16 |32 |
+-------------------------
+```
+
+And the sum: 1 + 2 + 4 + 8 + 16 + 32 = 63.  Hmm... I'm actually starting to see a glimmer of a pattern, but we'll do one more just to be sure.
+
+7 squares:
+
+```txt
+-----------------------------
+| 1 | 2 | 4 | 8 |16 |32 |64 |
+-----------------------------
+```
+
+1 + 2 + 4 + 8 + 16 + 32 +64 = 127.  Do you see it?  Anything ringing any alarms with the values 31, 63, 127?
+
+They're *almost* powers of 2.  In fact, they're *one less* than the *next* power of two.  For another example: if the biggest square on your board was 4096, you can bet your bottom dollar that the total will be 8191.  (1 + 2 + 4 + 8 + 16 + 32 + 64 + 128 + 256 + 512 + 1024 + 2048 + 4096 = 8191).
```

Comment:
I totally see where you're going with this one... _For another example: if the biggest square on your board was 4096, you can bet your bottom dollar that the total will be 8191._. I'm wonder if this would help emphasize it:

> Now, imagine a 12-square chess board. That's twelve doublings, or 2 multiplied by itself 12 times, which in the math biz is 2<sup>12</sup>. Or, for us mere mortals, 4096. Double that again and you get 8192. So... if we got the pattern right, the running total would be _almost_ 8192, also known as 8191. And if we tally it up, that's, exactly what we get: 1 + 2 + 4 + 8 + 16 + 32 + 64 + 128 + 256 + 512 + 1024 + 2048 + 4096 = 8191.